File: src/tests/query_test3d.cpp
#include "tests/query_test3d.h"
#include "query_test3d.h"
#include <godot_cpp/core/class_db.hpp>

void QueryTest3D::perform_test(Ref<QueryItem3D> projection) {
	GDVIRTUAL_CALL(_perform_test, projection);
}

void QueryTest3D::_bind_methods() {
	GDVIRTUAL_BIND(_perform_test, "projection");
	ClassDB::bind_method(D_METHOD("set_context_score_operator", "score_op"), &QueryTest3D::set_context_score_operator);
	ClassDB::bind_method(D_METHOD("get_context_score_operator"), &QueryTest3D::get_context_score_operator);

	ClassDB::bind_method(D_METHOD("set_context_filter_operator", "score_op"), &QueryTest3D::set_context_filter_operator);
	ClassDB::bind_method(D_METHOD("get_context_filter_operator"), &QueryTest3D::get_context_filter_operator);

	ClassDB::bind_method(D_METHOD("set_test_type", "type"), &QueryTest3D::set_test_type);
	ClassDB::bind_method(D_METHOD("get_test_type"), &QueryTest3D::get_test_type);

	ClassDB::bind_method(D_METHOD("set_test_purpose", "purpose"), &QueryTest3D::set_test_purpose);
	ClassDB::bind_method(D_METHOD("get_test_purpose"), &QueryTest3D::get_test_purpose);

	ClassDB::bind_method(D_METHOD("set_cost", "new_cost"), &QueryTest3D::set_cost);
	ClassDB::bind_method(D_METHOD("get_cost"), &QueryTest3D::get_cost);

	ClassDB::bind_method(D_METHOD("set_filter_type", "type"), &QueryTest3D::set_filter_type);
	ClassDB::bind_method(D_METHOD("get_filter_type"), &QueryTest3D::get_filter_type);

	ClassDB::bind_method(D_METHOD("set_filter_min", "val"), &QueryTest3D::set_filter_min);
	ClassDB::bind_method(D_METHOD("get_filter_min"), &QueryTest3D::get_filter_min);

	ClassDB::bind_method(D_METHOD("set_filter_max", "val"), &QueryTest3D::set_filter_max);
	ClassDB::bind_method(D_METHOD("get_filter_max"), &QueryTest3D::get_filter_max);

	ClassDB::bind_method(D_METHOD("set_scoring_factor", "val"), &QueryTest3D::set_scoring_factor);
	ClassDB::bind_method(D_METHOD("get_scoring_factor"), &QueryTest3D::get_scoring_factor);

	ClassDB::bind_method(D_METHOD("set_bool_match", "match"), &QueryTest3D::set_bool_match);
	ClassDB::bind_method(D_METHOD("get_bool_match"), &QueryTest3D::get_bool_match);

	BIND_ENUM_CONSTANT(FILTER_SCORE);
	BIND_ENUM_CONSTANT(FILTER_ONLY);
	BIND_ENUM_CONSTANT(SCORE_ONLY);

	BIND_ENUM_CONSTANT(TEST_TYPE_NUMERIC);
	BIND_ENUM_CONSTANT(TEST_TYPE_BOOLEAN);

	BIND_ENUM_CONSTANT(AVERAGE_SCORE);
	BIND_ENUM_CONSTANT(MAX_SCORE);
	BIND_ENUM_CONSTANT(MIN_SCORE);

	BIND_ENUM_CONSTANT(ANY_PASS);
	BIND_ENUM_CONSTANT(ALL_PASS);

	BIND_ENUM_CONSTANT(FILTER_TYPE_MIN);
	BIND_ENUM_CONSTANT(FILTER_TYPE_MAX);
	BIND_ENUM_CONSTANT(FILTER_TYPE_RANGE);

	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "cost"), "set_cost", "get_cost");
	ADD_PROPERTY(PropertyInfo(Variant::INT, "test_purpose", PROPERTY_HINT_ENUM, "Filter Score,Filter Only,Score Only"), "set_test_purpose", "get_test_purpose");
	ADD_PROPERTY(PropertyInfo(Variant::INT, "test_type", PROPERTY_HINT_ENUM, "Numeric,Boolean", PROPERTY_USAGE_STORAGE), "set_test_type", "get_test_type");
	ADD_PROPERTY(PropertyInfo(Variant::INT, "multiple_context_filter_operator", PROPERTY_HINT_ENUM, "Any Pass,All Pass"), "set_context_filter_operator", "get_context_filter_operator");
	ADD_PROPERTY(PropertyInfo(Variant::INT, "multiple_context_score_operator", PROPERTY_HINT_ENUM, "Average Score,Max Score,Min Score"), "set_context_score_operator", "get_context_score_operator");
	ADD_PROPERTY(PropertyInfo(Variant::INT, "filter_type", PROPERTY_HINT_ENUM, "Min,Max,Range"), "set_filter_type", "get_filter_type");
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "filter_min"), "set_filter_min", "get_filter_min");
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "filter_max"), "set_filter_max", "get_filter_max");
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "scoring_factor"), "set_scoring_factor", "get_scoring_factor");
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "bool_match"), "set_bool_match", "get_bool_match");
}

void QueryTest3D::_validate_property(PropertyInfo &property) const {
	// Hide filter operator when not filtering
	if (property.name == StringName("multiple_context_filter_operator"))
		if (get_test_purpose() == SCORE_ONLY)
			property.usage &= ~PROPERTY_USAGE_EDITOR;

	// Hide score operator when not scoring
	if (property.name == StringName("multiple_context_score_operator"))
		if (get_test_purpose() == FILTER_ONLY)
			property.usage &= ~PROPERTY_USAGE_EDITOR;

	// Hide numeric fields when this is a boolean test
	TypedArray<StringName> numeric_vars = { "filter_type", "filter_min", "filter_max", "scoring_factor" };
	if (numeric_vars.has(property.name))
		if (get_test_type() == TEST_TYPE_BOOLEAN)
			property.usage &= ~PROPERTY_USAGE_EDITOR;

	// Hide bool_match when this is a numeric test
	if (property.name == StringName("bool_match"))
		if (get_test_type() == TEST_TYPE_NUMERIC)
			property.usage &= ~PROPERTY_USAGE_EDITOR;

	// Hide min/max based on filter type
	if (property.name == StringName("filter_min"))
		if (get_filter_type() == FILTER_TYPE_MAX)
			property.usage &= ~PROPERTY_USAGE_EDITOR;

	if (property.name == StringName("filter_max"))
		if (get_filter_type() == FILTER_TYPE_MIN)
			property.usage &= ~PROPERTY_USAGE_EDITOR;
}

File: project/addons/geqo/csharp_binds/QueryTest3D.cs
using Godot;


///<summary>
/// Tests each QueryItem3D from a QueryGenerator3D to determine which is the best option.
/// Named the same as native QueryItem3D so the EnvironmentQuery is able to call it.
///</summary> 
[Tool]
public partial class QueryTest3D : Node3D
{
    public enum TestPurposeEnum { FilterScore = 0, FilterOnly = 1, ScoreOnly = 2 }
    public enum TestTypeEnum { Numeric = 0, Boolean = 1 }
    public enum FilterTypeEnum { Min = 0, Max = 1, Range = 2 }
    public enum MultipleContextScoreOp { AverageScore = 0, MaxScore = 1, MinScore = 2 }
    public enum MultipleContextFilterOp { AnyPass = 0, AllPass = 1 }
    public enum ScoreClampType { None = 0, Value = 1, SameAsFilter = 2 }

    public float Cost
    {
        get => (float)Call("get_cost");
        set => Call("set_cost", value);
    }
    public FilterTypeEnum FilterType
    {
        get => (FilterTypeEnum)(int)Call("get_filter_type");
        set => Call("set_filter_type", (int)value);
    }

    public bool BoolMatch
    {
        get => (bool)Call("get_bool_match");
        set => Call("set_bool_match", value);
    }

    public float FilterMin
    {
        get => (float)Call("get_filter_min");
        set => Call("set_filter_min", value);
    }

    public float FilterMax
    {
        get => (float)Call("get_filter_max");
        set => Call("set_filter_max", value);
    }

    public MultipleContextFilterOp MultipleContextFilterOperator
    {
        get => (MultipleContextFilterOp)(int)Call("get_multiple_context_filter_operator");
        set => Call("set_multiple_context_filter_operator", (int)value);
    }

    public MultipleContextScoreOp MultipleContextScoreOperator
    {
        get => (MultipleContextScoreOp)(int)Call("get_multiple_context_score_operator");
        set => Call("set_multiple_context_score_operator", (int)value);
    }

    public Curve ScoreCurve
    {
        get => (Curve)(GodotObject)Call("get_scoring_curve");
        set => Call("set_scoring_curve", value);
    }

    public ScoreClampType ScoreClampMinType
    {
        get => (ScoreClampType)(int)Call("get_clamp_min_type");
        set => Call("set_clamp_min_type", (int)value);
    }

    public float ScoreClampMin
    {
        get => (float)Call("get_score_clamp_min");
        set => Call("set_score_clamp_min", value);
    }

    public ScoreClampType ScoreClampMaxType
    {
        get => (ScoreClampType)(int)Call("get_clamp_max_type");
        set => Call("set_clamp_max_type", (int)value);
    }

    public float ScoreClampMax
    {
        get => (float)Call("get_score_clamp_max");
        set => Call("set_score_clamp_max", value);
    }

    public float ScoringFactor
    {
        get => (float)Call("get_scoring_factor");
        set => Call("set_scoring_factor", value);
    }

    public TestPurposeEnum TestPurpose
    {
        get => (TestPurposeEnum)(int)Call("get_test_purpose");
        set => Call("set_test_purpose", (int)value);
    }

    public TestTypeEnum TestType
    {
        get => (TestTypeEnum)(int)Call("get_test_type");
        set => Call("set_test_type", (int)value);
    }
    /// <summary>
    /// The GDExtension calls this function
    /// </summary>
    public void _perform_test(RefCounted queryInstance)
    {
        QueryInstanceWrapper3D instance = new QueryInstanceWrapper3D(queryInstance);
        _PerformTest(instance);
    }

    public virtual void _PerformTest(QueryInstanceWrapper3D queryInstance)
    {
        GD.PrintErr("Need to override PerformTest in derived class");
    }

    public void EndTest()
    {
        Call("end_test");
    }
}